src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers/ApiAlertAnalyzer.cs

```diff
@@ -0,0 +1,237 @@
+            // the expected full name. If the node resolves to a non-type symbol,
+            // return false as this could indicate a local variable or something similar
+            // with a name that matches the target.
+            if (symbol is not null)
+            {
+                if (symbol is ITypeSymbol typeSymbol)
+                {
+                    // This conditional can't be combined with the previous one because
+                    // failing this condition should not lead to the else clause.
+                    // In other words, if the symbol is a type symbol but is an error type
+                    // symbol then don't return here (rather than returning false as would
+                    // happen for non-type symbols).
+                    if (typeSymbol is not IErrorTypeSymbol)
+                    {
+                        return targetSyntax.NameMatcher.Matches(typeSymbol);
+                    }
+                }
+                else
+                {
+                    return false;
+                }
+            }
+
+            // If the node's full type can't be determined (either by symbol or fully qualified syntax),
+            // return true for the partial match only if ambiguous matching is enabled
+            return targetSyntax.AlertOnAmbiguousMatch;
+        }
+
+        private static bool AnalyzeMember(SyntaxNodeAnalysisContext context, TargetSyntax targetSyntax)
+        {
+            // If the node matches the target's fully qualified name, return true.
+            var qualifiedName = GetFullName(context.Node);
+            if (qualifiedName.Equals(targetSyntax.FullName, context.Node.GetStringComparison()))
+            {
+                return true;
+            }
+
+            // If the parent type's symbol is resolvable, return true if
+            // it corresponds to the target type
+            var typeSyntax = context.Node.Parent?.GetMAEExpressionSyntax();
+            if (typeSyntax is not null)
+            {
+                var symbol = context.SemanticModel.GetTypeInfo(typeSyntax).Type;
+                if (symbol is not null && symbol is not IErrorTypeSymbol)
+                {
+                    return targetSyntax.NameMatcher.Matches(symbol);
+                }
+            }
+
+            // If the node's full type can't be determined (either by symbol or fully qualified syntax),
+            // return true for the partial match only if ambiguous matching is enabled
+            return targetSyntax.AlertOnAmbiguousMatch;
+        }
+
+        private static string GetFullName(SyntaxNode node)
+        {
+            var fullName = node.GetQualifiedName().ToString();
+
+            // Ignore generic parameters for now to simplify matching
+            var genericParameterIndex = fullName.IndexOfAny(new[] { '<', '(' });
+            if (genericParameterIndex > 0)
+            {
+                fullName = fullName.Substring(0, genericParameterIndex);
```

Comment:
This pretty quickly returns as a string, though, so we'd still have to allocate. We could have the method return an ROS instead of a string, but that gets to be pretty viral. I'll leave this as-is for now, but ping me when you're back if you think this should be changed and we can re-visit.